keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,955 @@
+performance reasons, loggers like zap and klog do not filter out duplicates)
+and causes additional overhead.
+
+The guiding principle is that the callee should only add key/value pairs for
+information that the caller doesn’t have, like a local variable, and leave the
+decision about adding information about the parameters to the caller.
+
+Exceptions are when a log message is related to a parameter or a part of it
+that normally doesn’t need to be logged.
+
+This needs to be checked by taking a holistic view and ideally documenting the
+logging behavior.
+
+This situation is similar to wrapping errors: whether an error message contains
+information about a parameter (for example, a path name) needs to be documented
+to avoid storing redundant information when the caller wraps an error that was
+created by the callee.
+
+Analysis of the JSON log files collected at a high log level during a Prow test
+run can be used to detect cases where redundant key/value pairs occur in
+practice. This is not going to be complete, but it doesn’t have to be because
+the additional overhead for redundant key/value pairs matters a lot less when
+the message gets emitted infrequently.
+
+#### Modifying the same variable in a loop
+
+Reusing variable names keeps the code readable and prevents errors because the
+variable that isn’t meant to be used will be shadowed. However, care must be
+taken to really create new variables. This is broken:
+
+```go
+func foo(logger klogr.Logger, objects ...string) {
+    for _, obj := range objects {
+        // BUG: logger accumulates different key/value pairs with the same key
+        logger = logger.WithValue(“obj”, obj)
+        doSomething(logger, obj)
+    }
+}
+```
+
+A new variable must be created with `:=` inside the loop:
+
+```go
+func foo(logger klogr.Logger, objects ...string) {
+    for _, obj := range objects {
+        // This logger variable shadows the function parameter.
+        logger := logger.WithValue(“obj”, obj)
+        doSomething(logger, obj)
+    }
+}
+```
+
+#### Unused return values
+
+This code looks like it adds a name, but isn’t doing it correctly:
+
+```go
+func foo(logger klogr.Logger, objects ...string) {
+        // BUG: WithName returns a new logger with the name,
+        // but that return value is not used.
+        logger.WithName(“foo”)
+        doSomething(logger)
+    }
+}
+```
+
+
+## Design Details
+
+klog currently provides a formatter for log messages, a global logger instance,
+and some helper code.
+
+<<[UNRESOLVED @pohly ]>>
+[Log sanitization](https://github.com/kubernetes/enhancements/issues/1753) is an
+alpha feature with unclear next steps. Unless there is sufficient interest in
+that feature, it will not be supported anymore.
+<<[/UNRESOLVED]>>
+
+A new package `k8s.io/klogr` will replace both `k8s.io/klog` and
+`k8s.io/klog/klogr`. It will be hosted under `staging` and thus get developed
+together with the code that uses it, without a stable API guarantee. It’s
```

Comment:
something like `klogr.KObj` could be useful for developers of third-party controllers. Could that be part of `client-go` instead?

Reply:
`klogr` will be usable by anyone, with no dependencies other than the Go runtime and `go-logr/logr`. IMHO it's a better place than `client-go` because it's smaller and the purpose is specifically for logging.

If the "no stable API guarantee" part becomes a problem, then once the code is mature we can also consider graduating it out of staging into a separate repo with independent releases and a `v1` API.